snapshot_creator.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2021 the v8go contributors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package v8go
+
+// #include <stdlib.h>
+// #include "v8go.h"
+import "C"
+import (
+	"errors"
+	"unsafe"
+)
+
+type FunctionCodeHandling int
+
+const (
+	FunctionCodeHandlingKlear FunctionCodeHandling = iota
+	FunctionCodeHandlingKeep
+)
+
+type StartupData struct {
+	data     []byte
+	raw_size C.int
+	index    C.size_t
+}
+
+type SnapshotCreator struct {
+	ptr   C.SnapshotCreatorPtr
+	iso   *Isolate
+	ctx   *Context
+	index C.size_t
+}
+
+func NewSnapshotCreator() *SnapshotCreator {
+	v8once.Do(func() {
+		C.Init()
+	})
+
+	rtn := C.NewSnapshotCreator()
+
+	return &SnapshotCreator{
+		ptr: rtn.creator,
+		iso: &Isolate{ptr: rtn.iso},
+	}
+}
+
+func (s *SnapshotCreator) GetIsolate() (*Isolate, error) {
+	if s.ptr == nil {
+		return nil, errors.New("v8go: Cannot get Isolate after creating the blob")
+	}
+
+	return s.iso, nil
+}
+
+func (s *SnapshotCreator) AddContext(ctx *Context) error {
+	if s.ptr == nil {
+		return errors.New("v8go: Cannot add context to snapshot creator after creating the blob")
+	}
+
+	s.index = C.AddContext(s.ptr, ctx.ptr)
+	s.ctx = ctx
+
+	return nil
+}
+
+func (s *SnapshotCreator) Create(functionCode FunctionCodeHandling) (*StartupData, error) {
+	if s.ptr == nil {
+		return nil, errors.New("v8go: Cannot use snapshot creator after creating the blob")
+	}
+
+	if s.ctx == nil {
+		return nil, errors.New("v8go: Cannot create a snapshot without first adding a context")
+	}
+
+	rtn := C.CreateBlob(s.ptr, s.ctx.ptr, C.int(functionCode))
+
+	s.ptr = nil
+	s.ctx.ptr = nil
+	s.iso.ptr = nil
```

Comment:
So, the DCHECK in `~SnapshotCreator` seems to lock is into a requirement that we need to call CreateBlob on it before we can delete it.  

What if we made Create always call CreateBlob, and that calling Create was the only way to delete the snapshot creator (which again, is forced on us by the debug check).

Ugly, but lets us also remove `DeleteSnapshotCreator` and its caller:
```suggestion
	var ctxPtr C.ContextPtr // assuming this is a null *Context
	if s.ctx != nil {
		ctxPtr = s.ctx.ptr
	}

	rtn := C.CreateBlob(s.ptr, ctxPtr, C.int(functionCode))
	defer C.SnapshotBlobDelete(rtn)

	s.ptr = nil
	s.iso.ptr = nil
	if s.ctx == nil {
		return nil, errors.New("v8go: Cannot create a snapshot without first adding a context")
	} else {
		s.ctx.ptr = nil
	}
```